modin/core/dataframe/base/dataframe/utils.py

```diff
@@ -48,3 +53,71 @@ class JoinType(Enum):  # noqa: PR01
     LEFT = "left"
     RIGHT = "right"
     OUTER = "outer"
+
+
+def join_columns(
+    left: pandas.Index,
+    right: pandas.Index,
+    left_on: Union[Iterable[IndexLabel], IndexLabel],
+    right_on: Union[Iterable[IndexLabel], IndexLabel],
+    suffixes: Tuple[str, str],
+) -> Tuple[pandas.Index, Dict[IndexLabel, IndexLabel], Dict[IndexLabel, IndexLabel]]:
+    """
+    Compute resulting columns for the two dataframes being merged.
+
+    Parameters
+    ----------
+    left : pandas.Index
+        Columns of the left frame to join.
+    right : pandas.Index
+        Columns of the right frame to join.
+    left_on : list-like or scalar
+        Column names on which the frames are joined in the left DataFrame.
+    right_on : list-like or scalar
+        Column names on which the frames are joined in the right DataFrame.
+    suffixes : tuple[str, str]
+        A 2-length sequence containing suffixes to append to the intersected columns.
+
+    Returns
+    -------
+    pandas.Index, dict[IndexLabel -> IndexLabel], dict[IndexLabel -> IndexLabel]
+        Returns columns for the resulting frame and mappings of old to new column
+        names for `left` and `right` accordingly.
+    """
+    left_on = cast(Iterable[IndexLabel], [left_on] if is_scalar(left_on) else left_on)
+    right_on = cast(
+        Iterable[IndexLabel], [right_on] if is_scalar(right_on) else right_on
+    )
+
+    left_conflicts = set(left) & (set(right) - set(right_on))  # type: ignore # set() doesn't understand IndexLabel
+    right_conflicts = set(right) & (set(left) - set(left_on))  # type: ignore # set() doesn't understand IndexLabel
+    conflicting_cols = left_conflicts | right_conflicts
+
+    def _get_new_name(col: IndexLabel, suffix: str) -> IndexLabel:
+        if col in conflicting_cols:
+            return (
+                (f"{col[0]}{suffix}", *col[1:])
+                if isinstance(col, tuple)
+                else f"{col}{suffix}"
+            )
+        else:
+            return col
+
+    left_renamer: Dict[IndexLabel, IndexLabel] = {}
+    right_renamer: Dict[IndexLabel, IndexLabel] = {}
+    new_left: List[IndexLabel] = []
+    new_right: List[IndexLabel] = []
```

Comment:
`IndexLabel = Union[Hashable, Sequence[Hashable]]` thus this type is defined in pandas.

Why not like this?
```suggestion
    new_left: IndexLabel = []
    new_right: IndexLabel = []
```

So far, it looks like type hints only complicate things.

Reply:
I have to use `List`, otherwise, it wouldn't allow me to use list's API (like `.append` and `+`) further on the code

Reply:
This will be a mistake. Is it possible to give some hints for custom types? How does pandas deal with it? As a last resort, we can turn off type checking or not use types from pandas.

Reply:
now using an untyped list here
```python
new_left: List = []
```